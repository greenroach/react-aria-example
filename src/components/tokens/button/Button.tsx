import { FC, ReactNode } from "react";
import {
  Button as AriaButton,
  ButtonProps as AriaButtonProps,
} from "react-aria-components";
import cx from "classnames";
import styles from "./Button.module.scss";

type Size = "small" | "medium";
type Variant = "contained" | "outlined";
type Theme = "primary" | "secondary";
type IconPosition = "left" | "right";

interface ButtonProps {
  size?: Size;
  variant?: Variant;
  theme?: Theme;
  disabled?: boolean;
  icon?: ReactNode;
  iconPosition?: IconPosition;
  children?: ReactNode;
  onPress?: AriaButtonProps["onPress"];
}

export const Button: FC<ButtonProps> = ({
  size = "medium",
  variant = "contained",
  theme = "primary",
  disabled,
  icon,
  iconPosition = "left",
  onPress,
  children,
}) => {
  const isIconOnly = !!icon && !children;

  const classes = cx(
    styles.button,
    styles[size],
    styles[variant],
    styles[theme],
    disabled && styles.disabled,
    isIconOnly && styles.iconOnly,
    !!icon &&
      !isIconOnly &&
      (iconPosition === "right" ? styles.iconRight : styles.iconLeft),
  );

  return (
    <AriaButton onPress={onPress} isDisabled={disabled} className={classes}>
      {icon && iconPosition === "left" && (
        <span className={styles.icon}>{icon}</span>
      )}
      {children && <span className={styles.label}>{children}</span>}
      {icon && iconPosition === "right" && (
        <span className={styles.icon}>{icon}</span>
      )}
    </AriaButton>
  );
};
