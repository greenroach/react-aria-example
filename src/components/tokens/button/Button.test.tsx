import "@testing-library/jest-dom";
import { render, screen, cleanup, waitFor } from "@testing-library/react";
import { describe, it, expect, vi, afterEach } from "vitest";
import { Button } from "./Button";
import { userEvent } from "@storybook/test";

const sizes = ["small", "medium"];
const variants = ["contained", "outlined"];
const themes = ["primary", "secondary"];

const generateButtonTestCases = () => {
  const testCases = [];

  for (const size of sizes) {
    for (const variant of variants) {
      for (const theme of themes) {
        testCases.push({ size, variant, theme });
      }
    }
  }

  return testCases;
};

describe("Button Component", () => {
  afterEach(() => {
    cleanup();
  });

  it("renders with default props", () => {
    render(<Button>Click Me</Button>);
    const button = screen.getByText("Click Me");

    expect(button).toMatchInlineSnapshot(`
      <span
        class="_label_40c60c"
      >
        Click Me
      </span>
    `);
  });
  const testCases = generateButtonTestCases();

  it.each(testCases)(
    "renders with size: $size, variant: $variant, theme: $theme",
    ({ size, variant, theme }) => {
      render(
        // @ts-expect-error passing raw strings from testcases
        <Button size={size} variant={variant} theme={theme}>
          "text"
        </Button>,
      );
      const button = screen.getByRole("button");
      expect(button).toHaveTextContent("text");
      expect(button).toMatchSnapshot();
    },
  );

  it("renders with an icon", () => {
    render(<Button icon={<span>Icon</span>}>Button with Icon</Button>);
    const icon = screen.getByText("Icon");
    expect(icon).toBeVisible();
  });

  it("handles the disabled state", () => {
    render(<Button isDisabled={true}>Disabled Button</Button>);
    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
  });

  it("calls onPress when clicked", async () => {
    const onPress = vi.fn();
    render(<Button onPress={onPress}>Clickable Button</Button>);
    const button = screen.getByRole("button");
    userEvent.click(button);

    await waitFor(() => {
      expect(onPress).toHaveBeenCalledTimes(1);
    });
  });

  it("does not accept style prop", async () => {
    //@ts-expect-error testing a wrong prop
    render(<Button style={{ display: "block" }}>Clickable Button</Button>);
    const button = screen.getByRole("button");
    expect(button).not.toHaveStyle({ display: "block" });
  });

  it("does not accept className prop", async () => {
    //@ts-expect-error testing a wrong prop
    render(<Button className="test-class-name">Clickable Button</Button>);
    const button = screen.getByRole("button");
    expect(button).not.toHaveClass("test-class-name");
  });
});
