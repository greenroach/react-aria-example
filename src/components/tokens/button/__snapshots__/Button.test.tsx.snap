// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > renders with size: 'medium', variant: 'contained', theme: 'primary' 1`] = `
<button
  class="_button_40c60c _medium_40c60c _contained_40c60c _primary_40c60c"
  data-rac=""
  id="react-aria-«ra»"
  tabindex="0"
  type="button"
>
  <span
    class="_label_40c60c"
  >
    "text"
  </span>
</button>
`;

exports[`Button Component > renders with size: 'medium', variant: 'contained', theme: 'secondary' 1`] = `
<button
  class="_button_40c60c _medium_40c60c _contained_40c60c _secondary_40c60c"
  data-rac=""
  id="react-aria-«rc»"
  tabindex="0"
  type="button"
>
  <span
    class="_label_40c60c"
  >
    "text"
  </span>
</button>
`;

exports[`Button Component > renders with size: 'medium', variant: 'outlined', theme: 'primary' 1`] = `
<button
  class="_button_40c60c _medium_40c60c _outlined_40c60c _primary_40c60c"
  data-rac=""
  id="react-aria-«re»"
  tabindex="0"
  type="button"
>
  <span
    class="_label_40c60c"
  >
    "text"
  </span>
</button>
`;

exports[`Button Component > renders with size: 'medium', variant: 'outlined', theme: 'secondary' 1`] = `
<button
  class="_button_40c60c _medium_40c60c _outlined_40c60c _secondary_40c60c"
  data-rac=""
  id="react-aria-«rg»"
  tabindex="0"
  type="button"
>
  <span
    class="_label_40c60c"
  >
    "text"
  </span>
</button>
`;

exports[`Button Component > renders with size: 'small', variant: 'contained', theme: 'primary' 1`] = `
<button
  class="_button_40c60c _small_40c60c _contained_40c60c _primary_40c60c"
  data-rac=""
  id="react-aria-«r2»"
  tabindex="0"
  type="button"
>
  <span
    class="_label_40c60c"
  >
    "text"
  </span>
</button>
`;

exports[`Button Component > renders with size: 'small', variant: 'contained', theme: 'secondary' 1`] = `
<button
  class="_button_40c60c _small_40c60c _contained_40c60c _secondary_40c60c"
  data-rac=""
  id="react-aria-«r4»"
  tabindex="0"
  type="button"
>
  <span
    class="_label_40c60c"
  >
    "text"
  </span>
</button>
`;

exports[`Button Component > renders with size: 'small', variant: 'outlined', theme: 'primary' 1`] = `
<button
  class="_button_40c60c _small_40c60c _outlined_40c60c _primary_40c60c"
  data-rac=""
  id="react-aria-«r6»"
  tabindex="0"
  type="button"
>
  <span
    class="_label_40c60c"
  >
    "text"
  </span>
</button>
`;

exports[`Button Component > renders with size: 'small', variant: 'outlined', theme: 'secondary' 1`] = `
<button
  class="_button_40c60c _small_40c60c _outlined_40c60c _secondary_40c60c"
  data-rac=""
  id="react-aria-«r8»"
  tabindex="0"
  type="button"
>
  <span
    class="_label_40c60c"
  >
    "text"
  </span>
</button>
`;
